src/main/java/com/redhat/rhjmc/containerjfr/net/web/http/api/v2/CertificatePostHandler.java

```diff
@@ -0,0 +1,169 @@
+import com.redhat.rhjmc.containerjfr.net.web.http.api.ApiVersion;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.ext.web.FileUpload;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.handler.impl.HttpStatusException;
+
+class CertificatePostHandler extends AbstractAuthenticatedRequestHandler {
+
+    static final String PATH = "certificates";
+
+    private final Environment env;
+    private final FileSystem fs;
+    private final Logger logger;
+
+    private static final String TRUSTSTORE_DIR = "TRUSTSTORE_DIR";
+
+    private Function<File, FileOutputStream> outputStreamFunction;
+    private CertificateValidator certValidator;
+
+    @Inject
+    CertificatePostHandler(
+            AuthManager auth,
+            Environment env,
+            FileSystem fs,
+            Logger logger,
+            @Named("OutputStreamFunction") Function<File, FileOutputStream> outputStreamFunction,
+            CertificateValidator certValidator) {
+        super(auth);
+        this.env = env;
+        this.fs = fs;
+        this.logger = logger;
+        this.outputStreamFunction = outputStreamFunction;
+        this.certValidator = certValidator;
+    }
+
+    @Override
+    public ApiVersion apiVersion() {
+        return ApiVersion.V2;
+    }
+
+    @Override
+    public HttpMethod httpMethod() {
+        return HttpMethod.POST;
+    }
+
+    @Override
+    public String path() {
+        return basePath() + PATH;
+    }
+
+    @Override
+    public boolean isAsync() {
+        return false;
+    }
+
+    @Override
+    public boolean isOrdered() {
+        return true;
+    }
+
+    @Override
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
+    public void handleAuthenticated(RoutingContext ctx) throws Exception {
+        FileUpload cert = null;
+        for (FileUpload fu : ctx.fileUploads()) {
+            if ("cert".equals(fu.name())) {
+                cert = fu;
+                break;
+            }
+        }
+
+        if (cert == null) {
+            throw new HttpStatusException(
+                    400, "A file named \"cert\" was not included in the request");
+        }
+
+        Path certPath = fs.pathOf(cert.uploadedFileName());
+
+        if (!env.hasEnv(TRUSTSTORE_DIR)) {
+            throw new HttpStatusException(500, "Truststore directory not set");
+        }
+
+        String truststoreDir = env.getEnv(TRUSTSTORE_DIR);
+        Path filePath = fs.pathOf(truststoreDir, cert.fileName()).normalize();
+
+        if (fs.exists(filePath)) {
+            throw new HttpStatusException(409, filePath.toString() + " Certificate already exists");
+        }
+
+        try (InputStream fis = fs.newInputStream(certPath);
+                DataInputStream dis = new DataInputStream(fis);
+                FileOutputStream out = outputStreamFunction.apply(filePath.toFile())) {
+
+            byte[] bytes = new byte[dis.available()];
```

Comment:
Are you sure `dis.available()` tells you the total size of the certificate data in all scenarios? Not all streams implement `available()` in this way - it's more common that this method reports the number of bytes that can be read without blocking on the next invocation of `read()`. If this is the case, then for larger certificates (or various other conditions), this byte buffer will end up undersized and only some initial chunk of the certificate will be read into the buffer, rather than the whole thing.

Why is the `ByteArrayInputStream` wrapping around this buffer needed to begin with? Can't `fis` or `dis` be passed directly to `parseCertificate`? The docs for `generateCertificate` say it accepts a generic `InputStream`, not specifically a `ByteArrayInputStream`.

Reply:
It gives parsing errors when passing in the InputStream or DataInputStream. However I could add the byte array part into the parseCertificate method.

Reply:
What "parsing errors"? Does it throw a `CertificateException`, and if so, does the exception say anything informative?

The official docs suggest that passing the `FileInputStream` wrapped in a `BufferedInputStream` should be sufficient: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/cert/CertificateFactory.html

ie.

```
 FileInputStream fis = new FileInputStream(filename);
 BufferedInputStream bis = new BufferedInputStream(fis);

 CertificateFactory cf = CertificateFactory.getInstance("X.509");

 while (bis.available() > 0) {
    Certificate cert = cf.generateCertificate(bis);
    System.out.println(cert.toString());
 }
```

and

```
 FileInputStream fis = new FileInputStream(filename);
 CertificateFactory cf = CertificateFactory.getInstance("X.509");
 Collection c = cf.generateCertificates(fis);
 Iterator i = c.iterator();
 while (i.hasNext()) {
    Certificate cert = (Certificate)i.next();
    System.out.println(cert);
 }
```

It would be much better if we can figure a way to generate the certificate more directly from a stream of the file, without first reading the whole file into an application memory buffer and making a stream out of that. Less bug prone and probably better overall throughput performance.

Reply:
I changed the certificateValidator, it now takes in a InputStream, and outputs a Collection.